### ISSUE TITLE
[QUESTION]mydumper data lost for bigtable

### ISSUE THREAD

**Issue by shoshirakawa** [mydumper/mydumper#1337]
**Describe the bug**
data lost for bigtable( record > 19,000,000) in mydumper

**To Reproduce**
Command executed:
time mydumper --threads=16 --user=root --password=password --host=localhost --port=13306 --protocol=tcp \
--compress=ZSTD --outputdir=bigtable --tables-list=sample.bigtable \
--routines --triggers --events --rows=100000 --verbose=3 --logfile=mydumper-bigtable.log --trx-consistency-only

What mydumper and myloader version has been used?
mydumper v0.15.1-3

**Expected behavior**
real data :                   select count(1) from bigtable ->   19175087
result in metadata:    real_table_name=bigtable rows = 19175084
 and i imported the exported data,  just  19175084 records in my new table.

**Log**
If applicable, add the --verbose 3 / -v 3 and --logfile / -L <filename> to the execution of the command and upload the file.
[issues.log](https://github.com/mydumper/mydumper/files/13222442/issues.log)

**Backup**
i use gcp cloudsql mysql 5.7.43-google-log (Google) i can not log in the real server.

**How to repeat**
If applicable, add the minimal table structure and data that we need to reproduce the issue.
Or upload a [core dump](https://github.com/mydumper/mydumper/wiki/Support#core-dump) for cases where mydumper or myloader is crashing.

The program did not crash. no coredump.
my table structure:
+---------------+------+-----+---------+-------+
| Type          | Null | Key | Default | Extra |
+---------------+------+-----+---------+-------+
| int(11)       | NO   | PRI | NULL    |       |
| int(11)       | NO   | MUL | NULL    |       |
| int(11)       | NO   |     | NULL    |       |
| int(11)       | NO   | MUL | NULL    |       |
| int(11)       | YES  | MUL | NULL    |       |
| int(11)       | NO   |     | NULL    |       |
| int(11)       | YES  | MUL | NULL    |       |
| int(11)       | NO   | MUL | NULL    |       |
| varchar(255)  | YES  |     | NULL    |       |
| int(11)       | YES  |     | NULL    |       |
| tinyint(2)    | NO   |     | 0       |       |
| float         | NO   |     | NULL    |       |
| int(11)       | NO   |     | NULL    |       |
| float         | NO   |     | NULL    |       |
| float         | YES  |     | NULL    |       |
| varchar(2048) | NO   |     |         |       |
| tinyint(4)    | NO   |     | 1       |       |
| json          | YES  |     | NULL    |       |
| json          | YES  |     | NULL    |       |
| datetime      | NO   | MUL | NULL    |       |
| date          | NO   | MUL | NULL    |       |
| datetime      | YES  |     | NULL    |       |

**Environment (please complete the following information):**
 - OS version: [e.g. CentOS 7.9, Ubuntu 20.04]
 - MyDumper version: [ mydumper --version ]
 
DISTRIB_DESCRIPTION="Ubuntu 22.04.3 LTS"
mydumper v0.15.1-3, built against MySQL 8.0.33-25 with SSL support

**Additional context**

when i change rows from 100000 to 200000, no problem.
19175087 records have been exported and imported to new table
so it is a Randomly occurring issues. 

thanks.

**Comment by midenok** [comment]
First of all, why did you add ASCII art table? How to reproduce with that? Please use `SHOW CREATE TABLE`.

@shoshirakawa Can you confirm this is myloader or mydumper issue? You should count INSERT statements in the produced dump.

Also if it is stochastic you can compile latest git revision with -DWITH_ASAN=ON and check if it fails. AFAIU there was many changes since 0.15.1, if it is not reproducible on latest revision this bug should be reported into the corresponding distribution.

**Comment by shoshirakawa** [comment]
@midenok 
thank you for your reply.

1. 
> First of all, why did you add ASCII art table? How to reproduce with that? Please use SHOW CREATE TABLE.

it is not a ascii art table, just desc bigtable and remove the column name for security reason. sry.
this is the result of `show create table`: 

CREATE TABLE `bigtable` (
  `id` int(11) NOT NULL,
  `cs01_id` int(11) NOT NULL,
  `cv_id` int(11) NOT NULL,
  `othertable01_id` int(11) NOT NULL,
  `wb_deliver_id` int(11) DEFAULT NULL,
  `original_id` int(11) NOT NULL,
  `root_structure_element_id` int(11) DEFAULT NULL,
  `cses_id` int(11) NOT NULL,
  `cses_device_code` varchar(255) DEFAULT NULL,
  `review_target_cses_id` int(11) DEFAULT NULL,
  `correct` tinyint(2) NOT NULL DEFAULT '0' COMMENT '0->incorrect, 1->correct',
  `score` float NOT NULL,
  `redo_count` int(11) NOT NULL,
  `time` float NOT NULL,
  `read_asr_time` float DEFAULT NULL,
  `asr_text` varchar(2048) NOT NULL DEFAULT '',
  `asr_text_type` tinyint(4) NOT NULL DEFAULT '1',
  `play_situation` json DEFAULT NULL,
  `attributes` json DEFAULT NULL,
  `asred_at` datetime NOT NULL,
  `asred_date` date NOT NULL,
  `qubena_log_migrated_at` datetime DEFAULT NULL,
  PRIMARY KEY (`id`),
  KEY `fk_bigtable_othertable011` (`othertable01_id`),
  KEY `fk_bigtable_wb_deliver1` (`wb_deliver_id`),
  KEY `fk_bigtable_lcs1` (`cses_id`),
  KEY `idx_asred_date_asr_log` (`asred_date`),
  KEY `idx_asred_at_asr_log` (`asred_at`),
  KEY `fk_bigtable_cs011_idx` (`cs01_id`,`asred_at`),
  KEY `fk_bigtable_root_structure_element_id_idx` (`root_structure_element_id`,`cs01_id`,`othertable01_id`),
  CONSTRAINT `fk_bigtable_cs011` FOREIGN KEY (`cs01_id`) REFERENCES `cs01` (`id`),
  CONSTRAINT `fk_bigtable_lcs1` FOREIGN KEY (`cses_id`) REFERENCES `lcs` (`id`),
  CONSTRAINT `fk_bigtable_othertable011` FOREIGN KEY (`othertable01_id`) REFERENCES `othertable01` (`id`),
  CONSTRAINT `fk_bigtable_wb_deliver1` FOREIGN KEY (`wb_deliver_id`) REFERENCES `wb_deliver` (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4

2.  
> Can you confirm this is myloader or mydumper issue? You should count INSERT statements in the produced dump.
   yes. it is a mydumper issue. i have counted insert statements in the dump. = 19175084 (not a correct result).
   
 cat sample.bigtable.*.sql | egrep "^\,\(" | wc -l
  19170634
cat sample.bigtable.*.sql | grep INSERT | wc -l
  4450
echo "4450 + 19170634" | bc
  19175084

3. 
> Also if it is stochastic you can compile latest git revision with -DWITH_ASAN=ON and check if it fails. AFAIU there was many changes since 0.15.1, if it is not reproducible on latest revision this bug should be reported into the corresponding distribution.

     thank you for your advice. i will try it and feedback the result.

**Comment by shoshirakawa** [comment]
@midenok 

1. 
i cloned mydumper master and compiled with  -DWITH_ASAN=ON option, but failed. 
   make error log
[make_error.log](https://github.com/mydumper/mydumper/files/13233534/make_error.log)

2.
good news. 
I compiled the program with the -DWITH_ASAN=ON option and used it to export data. 
As a result, I found that the BUG had been fixed.

Using the same data for testing

in mydumper v0.15.1-3
I executed a total of 20 data export operations and recorded the outcomes.
The result was 4 failures. (4failure/20total = 20%)

However, use newly built mydumper (git hash: d84aac451a07413802ccbebd286abe54f3d1062b)
 I executed 40 times data export operations,
all results were correct.
 
 @davidducos 
 thank you too. a great tool for me!
